src/main/scala/bozzy/steno/StenoLayout.scala

```diff
@@ -1,25 +1,152 @@
+    beginningKeys foreach(character => {
+      builder ++= StenoLayout.characterToEscapedString(character)
+      builder += '?'
+    })
+    builder ++= ")?("
+    centerKeys._1 foreach(character => {
+      builder ++= StenoLayout.characterToEscapedString(character)
+      builder += '?'
+    })
+    builder ++= ")?("
+    endingKeys foreach(character => {
+      builder ++= StenoLayout.characterToEscapedString(character)
+      builder += '?'
+    })
+    builder ++= ")?$"
+    builder.toString.r
+  }
+}
 
 object StenoLayout {
   val STANDARD_STENO_LAYOUT =
     new StenoLayout( "English Stenography"
                    , "STKPWHR"
-                   , ("AO*EU", "-")
-                   , "FRPBLGTSZD"
+                   , ("AO-*EU", "-")
+                   , "FRPBLGTSDZ"
                    )
   val STANDARD_NUMBER_LAYOUT =
     new StenoLayout( "English Stenography Number Bar"
                    , "12K3W4R"
-                   , ("50*EU", "-")
+                   , ("50-*EU", "-")
                    , "6R7B8G9SDZ"
+                   , "#"
                    )
+  val numeric = """\d""".r
+
+  /** Normalize a set of chords, separated by the '/' character.
+    *
+    * Parses a stroke; a set of chords; into a consistent format
+    * and returns whether or not the stroke is valid syntax according
+    * to the steno layout.
+    */
+  def normalizeStroke(stroke: String) = {
+    val chords = (stroke split '/')
+    if (stroke.charAt(0) == '/' || stroke.last == '/') {
+      (stroke, false)
+    } else {
+      def assembleChords(remaining: Array[String], result: StringBuilder): (String, Boolean) = {
+        val (chord, valid) = normalizeChord(remaining.head)
+        if (!valid) {
+          (stroke, valid)
+        } else if (remaining.tail.size == 0) {
+          result ++= chord
+          (result.toString, valid)
+        } else {
+          result ++= s"$chord/"
+          assembleChords(remaining.tail, result)
+        }
+      }
+      assembleChords(chords, new StringBuilder(stroke.length))
+    }
+  }
+
+  /** Normalize a single chord to a unified format, and validate it.
+    *
+    * Parses a single chord, then spits out an equivalent chord in a
+    * consistent style, meaning that we can use it to compare
+    * identical chords that are formatted differently. We also validate
+    * the chord according to steno order and keys, and if the parsing
+    * fails, we return the given chord and valid false.
+    */
+  def normalizeChord(chord: String) = {
+    val result = new StringBuilder(chord.length * 2)
+    var valid = true // Will only return "normalized steno" if valid.
+
+    // Get right pattern if numeric or not.
+    val (pattern, cleaned) =
+      if (chord.contains(STANDARD_NUMBER_LAYOUT.flag) || // Contains #.
+        numeric.findFirstIn(chord).isDefined) { // Contains number.
+      val noHash = chord.replaceAllLiterally("#", "") // Get rid of #.
+      result += '#' // Result will start with #.
```

Comment:
Maybe I should drop the pound if the stroke is numeric... I'll look at some other dictionaries.